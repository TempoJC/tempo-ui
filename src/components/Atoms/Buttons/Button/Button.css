.tempo-button {
  border: var(--tempo-border-width) solid var(--color-token-light-semantic-highlight);
  outline: 0 none;
  display: flex;
  align-items: center;
  justify-content: center;
  padding: calc(15px - var(--tempo-border-width)) var(--size7);
  white-space: nowrap;
  font: var(--typography-label-l-regular-font);
  letter-spacing: var(--typography-label-l-regular-letter-spacing);
  text-transform: var(--typography-label-l-regular-text-transform);
  position: relative;
  max-width: 100%;
}

.tempo-button--extra {
  padding: calc(19px - var(--tempo-border-width)) var(--size7);
}

.tempo-button--icon {
  padding: calc(14px - var(--tempo-border-width));
}

.tempo-button--icon.tempo-button--extra {
  padding: calc(18px - var(--tempo-border-width));
}

.tempo-button--icon-and-label {
  padding: calc(14px - var(--tempo-border-width)) var(--size7);
}

.tempo-button--icon-and-label.tempo-button--extra {
  padding: calc(18px - var(--tempo-border-width)) var(--size7);
}

.tempo-button--dropdown {
  padding: calc(14px - var(--tempo-border-width));
  padding-left: var(--size6);
}

.tempo-button--dropdown.tempo-button--extra {
  padding: calc(18px - var(--tempo-border-width));
  padding-left: var(--size6);
}

@media (--viewport-l-xl) {
  .tempo-button {
    padding: calc(11px - var(--tempo-border-width)) var(--size7);
  }

  .tempo-button--extra {
    padding: calc(15px - var(--tempo-border-width)) var(--size7);
  }

  .tempo-button--icon {
    padding: calc(10px - var(--tempo-border-width));
  }

  .tempo-button--icon.tempo-button--extra {
    padding: calc(14px - var(--tempo-border-width));
  }

  .tempo-button--icon-and-label {
    padding: calc(10px - var(--tempo-border-width)) var(--size7);
  }

  .tempo-button--icon-and-label.tempo-button--extra {
    padding: calc(14px - var(--tempo-border-width)) var(--size7);
  }

  .tempo-button--dropdown {
    padding: calc(10px - var(--tempo-border-width));
    padding-left: var(--size5);
  }

  .tempo-button--dropdown.tempo-button--extra {
    padding: calc(14px - var(--tempo-border-width));
    padding-left: var(--size5);
  }
}

.tempo-button--standard {
  color: var(--color-token-light-content-1);
  background: transparent;
}
.tempo-button--color-dark.tempo-button--standard {
  color: var(--color-token-dark-content-1);
  border-color: var(--color-token-dark-semantic-highlight);
}

.tempo-button:focus-visible,
.tempo-button--loading:focus-visible {
  background: var(--color-token-light-semantic-highlight-2);
  color: var(--color-token-light-content-inverse);
}

/* Higher specificity needed to override colors depending on the button type */
.tempo-button.tempo-button--color-dark:focus-visible {
  background: var(--color-token-dark-semantic-highlight-2);
  color: var(--color-token-dark-content-1);
}

.tempo-button--standard:hover {
  cursor: pointer;
  color: var(--color-token-light-content-hover);
  border-color: var(--color-token-light-semantic-highlight-hover);
}
.tempo-button--color-dark.tempo-button--standard:hover {
  color: var(--color-token-dark-content-hover);
  border-color: var(--color-token-dark-semantic-highlight-hover);
}

.tempo-button--standard:active {
  color: var(--color-token-light-content-active);
  border-color: var(--color-token-light-semantic-highlight-active);
}
.tempo-button--color-dark.tempo-button--standard:active {
  color: var(--color-token-dark-content-active);
  border-color: var(--color-token-dark-semantic-highlight-active);
}

.tempo-button--standard:disabled {
  cursor: default;
  color: var(--color-token-light-content-disabled);
  border-color: var(--color-token-light-border-3);
}
.tempo-button--color-dark.tempo-button--standard:disabled {
  color: var(--color-token-dark-content-disabled);
  border-color: var(--color-token-dark-border-3);
}

.tempo-button--standard.tempo-button--loading {
  cursor: default;
  color: var(--color-token-light-content-disabled);
  border-color: var(--color-token-light-border-3);
}
.tempo-button--color-dark.tempo-button--standard.tempo-button--loading {
  color: var(--color-token-dark-content-disabled);
  border-color: var(--color-token-dark-border-3);
}

.tempo-button--primary {
  color: var(--color-token-light-content-inverse);
  background: var(--color-token-light-semantic-highlight);
  border-color: transparent;
}
.tempo-button--color-dark.tempo-button--primary {
  color: var(--color-token-dark-content-inverse);
  background: var(--color-token-dark-semantic-highlight);
}

.tempo-button--primary:hover {
  cursor: pointer;
  color: var(--color-token-light-content-inverse);
  background: var(--color-token-light-semantic-highlight-hover);
}
.tempo-button--color-dark.tempo-button--primary:hover {
  color: var(--color-token-dark-content-inverse);
  background: var(--color-token-dark-semantic-highlight-hover);
}

.tempo-button--primary:active {
  color: var(--color-token-light-content-inverse);
  background: var(--color-token-light-semantic-highlight-active);
}
.tempo-button--color-dark.tempo-button--primary:active {
  color: var(--color-token-dark-content-inverse);
  background: var(--color-token-dark-semantic-highlight-active);
}

.tempo-button--primary:disabled {
  cursor: default;
  color: var(--color-token-light-content-disabled);
  background: var(--color-token-light-background-disabled);
}

.tempo-button--color-dark.tempo-button--primary:disabled {
  color: var(--color-token-dark-content-disabled);
  background: var(--color-token-dark-background-disabled);
}

.tempo-button--primary.tempo-button--loading {
  cursor: default;
  color: var(--color-token-light-content-disabled);
  background: var(--color-token-light-background-disabled);
  border-color: var(--color-token-light-border-3);
}
.tempo-button--color-dark.tempo-button--primary.tempo-button--loading {
  color: var(--color-token-dark-content-disabled);
  background: var(--color-token-dark-background-disabled);
  border-color: var(--color-token-dark-border-3);
}

.tempo-button--danger {
  color: var(--color-token-light-content-inverse);
  background: var(--color-token-light-semantic-danger);
  border-color: transparent;
}
.tempo-button--color-dark.tempo-button--danger {
  color: var(--color-token-dark-content-inverse);
  background: var(--color-token-dark-semantic-danger);
}

.tempo-button--danger:hover {
  cursor: pointer;
  color: var(--color-token-light-content-inverse);
  background: var(--color-token-light-semantic-danger-hover);
}
.tempo-button--color-dark.tempo-button--danger:hover {
  color: var(--color-token-dark-content-inverse);
  background: var(--color-token-dark-semantic-danger-hover);
}

.tempo-button--danger:active {
  color: var(--color-token-light-content-inverse);
  background: var(--color-token-light-semantic-danger-active);
}
.tempo-button--color-dark.tempo-button--danger:active {
  color: var(--color-token-dark-content-inverse);
  background: var(--color-token-dark-semantic-danger-active);
}

.tempo-button--danger:disabled {
  cursor: default;
  color: var(--color-token-light-content-disabled);
  background: var(--color-token-light-background-disabled);
}
.tempo-button--color-dark.tempo-button--danger:disabled {
  color: var(--color-token-dark-content-disabled);
  background: var(--color-token-dark-background-disabled);
}

.tempo-button--danger.tempo-button--loading {
  cursor: default;
  color: var(--color-token-light-content-disabled);
  background: var(--color-token-light-background-disabled);
  border-color: var(--color-token-light-border-3);
}
.tempo-button--color-dark.tempo-button--danger.tempo-button--loading {
  color: var(--color-token-dark-content-disabled);
  background: var(--color-token-dark-background-disabled);
  border-color: var(--color-token-dark-border-3);
}

.tempo-button--dropdown .tempo-button__content > * + * {
  margin-left: var(--size1);
}

.tempo-button__content > * + * {
  margin-left: var(--size1);
}

.tempo-button--full-width {
  width: 100%;
  justify-content: center;
}

.tempo-button__label {
  overflow: hidden;
  text-overflow: ellipsis;
}

/* Specificity problems... */
.tempo-button .tempo-button__loader {
  position: absolute;
  left: 50%;
  top: 50%;
  transform: translate(-50%, -50%);
}

.tempo-button__content {
  display: inline-flex;
  align-items: center;
  max-width: 100%;
}

.tempo-button__content--loading {
  visibility: hidden;
}
