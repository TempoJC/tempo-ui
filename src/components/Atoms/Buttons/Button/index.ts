// packages/components-action/button
export { default as Button } from "./Button";
