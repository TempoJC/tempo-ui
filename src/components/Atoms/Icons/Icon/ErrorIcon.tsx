const ErrorIcon = () => (
  <svg role="img" width={20} height={20} viewBox="0 0 20 20" fill="none" xmlns="http://www.w3.org/2000/svg">
    <path
      fill="currentcolor"
      fillRule="evenodd"
      d="M10 17a7 7 0 1 0 0-14 7 7 0 0 0 0 14Zm0 1a8 8 0 1 0 0-16 8 8 0 0 0 0 16Z"
      clipRule="evenodd"
    />
  </svg>
);

export default ErrorIcon;
