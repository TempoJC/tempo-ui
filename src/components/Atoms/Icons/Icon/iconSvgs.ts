const svgModules = import.meta.glob("../Svgs/**/*.svg", { query: "?raw", import: "default" });

export const getIconSvg = (group: string, name: string): Promise<string> => {
  const path = `../Svgs/${group}/icon-${name}.svg`;
  const importFn = svgModules[path];

  if (!importFn) {
    return Promise.reject(new Error(`Icon not found at path: ${path}`));
  }

  return importFn() as Promise<string>;
};
