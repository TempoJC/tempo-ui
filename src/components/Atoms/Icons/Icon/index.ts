// packages/icons/amiga-icon
export { default as Icon } from "./Icon";
