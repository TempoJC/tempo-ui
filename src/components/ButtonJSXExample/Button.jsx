import React from "react";

const Button = ({ children, onClick }) => {
  return (
    <button
      onClick={onClick}
      style={{
        padding: "10px 20px",
        borderRadius: "5px",
        border: "1px solid #ccc",
        backgroundColor: "#f0f0f0",
        cursor: "pointer",
      }}
    >
      {children}
    </button>
  );
};

export default Button;
