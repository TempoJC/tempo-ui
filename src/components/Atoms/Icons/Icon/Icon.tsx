import React, { cloneElement } from "react";
import cn from "classnames";
import useInlineIcons from "./hooks/useInlineIcons";
import ErrorIcon from "./ErrorIcon";
import type { IconProps } from "./types";
import "./Icon.css";

const Icon: React.FC<IconProps> = ({ className, testId, group, name, errorFallback = <ErrorIcon />, ...rest }) => {
  const { error, svgRef } = useInlineIcons({ group, name });

  const computedClassName = cn("tempo-icon", className);

  const commonProps = {
    "data-icon-name": name,
    "data-icon-group": group,
    className: computedClassName,
    ...rest,
  };

  if (error) {
    return cloneElement(errorFallback, commonProps);
  }

  return (
    <svg
      ref={svgRef}
      role="img"
      width={20}
      height={20}
      viewBox="0 0 20 20"
      data-testid={testId}
      className={computedClassName}
      xmlns="http://www.w3.org/2000/svg"
      data-icon-group={group}
      data-icon-name={name}
      {...rest}
    />
  );
};

export default Icon;
