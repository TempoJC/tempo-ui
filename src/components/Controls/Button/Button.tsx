import { forwardRef } from "react";
import cn from "classnames";
import Loader from "@components/Loaders/Loader";
import Icon from "@components/Icons/Icon";
import type { ButtonProps } from "./types";
import "./Button.css";

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      className,
      onClick,
      onMouseDown,
      onMouseUp,
      onMouseLeave,
      onTouchStart,
      onTouchEnd,
      disabled = false,
      fullWidth = false,
      tabIndex,
      autoFocus = false,
      label,
      iconStart,
      iconEnd,
      testId,
      title,
      type = "button",
      kind = "standard",
      loading = false,
      extra = false,
      dropdown = false,
      dropdownVisible = false,
      color = "light",
      form,
    },
    ref,
  ) => {
    const hasIcon = iconStart || iconEnd;
    const iconAndLabel = hasIcon && label;
    const iconNoLabel = hasIcon && !label;

    return (
      <button
        ref={ref}
        className={cn(className, "tempo-button", {
          "tempo-button--color-dark": color === "dark",
          "tempo-button--full-width": fullWidth,
          "tempo-button--standard": kind === "standard",
          "tempo-button--primary": kind === "primary",
          "tempo-button--danger": kind === "danger",
          "tempo-button--extra": extra,
          "tempo-button--icon": iconNoLabel,
          "tempo-button--icon-and-label": iconAndLabel,
          "tempo-button--loading": loading,
          "tempo-button--dropdown": dropdown,
        })}
        onClick={onClick}
        onMouseDown={onMouseDown}
        onMouseUp={onMouseUp}
        onMouseLeave={onMouseLeave}
        onTouchStart={onTouchStart}
        onTouchEnd={onTouchEnd}
        disabled={disabled || loading}
        tabIndex={tabIndex}
        autoFocus={autoFocus}
        title={title}
        data-testid={testId}
        type={type}
        form={form}
      >
        {loading && <Loader size="small" color={color} className="tempo-button__loader" />}
        <div className={cn("tempo-button__content", { "tempo-button__content--loading": loading })}>
          {iconStart}
          {label && <span className={"tempo-button__label"}>{label}</span>}
          {iconEnd}
          {dropdown && dropdownVisible && <Icon group="arrows" name="chevron-up" />}
          {dropdown && !dropdownVisible && <Icon group="arrows" name="chevron-down" />}
        </div>
      </button>
    );
  },
);

export default Button;
