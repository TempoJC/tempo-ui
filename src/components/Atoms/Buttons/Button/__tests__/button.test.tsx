import { userEvent } from "@testing-library/user-event";
import { render, screen, waitFor } from "@testing-library/react";
import Button from "../Button";

// This is crucial to isolate Button's tests from Icon's internal logic,
// including its asynchronous SVG loading and any bundler-specific features like import.meta.glob.
jest.mock("@components/Atoms/Icons/Icon", () => ({
  // For a React component, Jest needs to mock the component function itself.
  // We use jest.fn() to create a mock function that will behave like a React component.
  Icon: jest.fn(({ group, name, className, ...props }) => {
    // Return a simple SVG element or a div to simulate the icon's presence.
    // It's important to pass relevant props as data-attributes to allow testing Button's usage.
    return (
      <svg
        data-testid={`mock-icon-${group}-${name}`} // Use a data-testid for easy lookup in tests
        data-icon-group={group} // Simulate attributes Icon would normally set
        data-icon-name={name} // Simulate attributes Icon would normally set
        className={className} // Pass the className if Button's styling depends on it
        {...props} // Pass any other props that Icon might receive and Button might forward
      />
    );
  }),
}));

// Similar to Icon, we mock Loader to prevent its internal logic from running
// and to isolate Button's tests from Loader's implementation details.
jest.mock("@components/Feedback/Loader", () => ({
  Loader: jest.fn(({ className, size, color, ...props }) => {
    // Return a simple div. It's important to:
    // 1. Have a data-testid for easy lookup.
    // 2. Apply the 'className' passed by Button, if the test verifies it.
    return (
      <div
        data-testid="mock-loader" // A generic data-testid for the mocked loader
        className={className} // Apply the className passed by Button (e.g., "tempo-button__loader")
        data-loader-size={size} // Optional: to verify if Button passes size correctly
        data-loader-color={color} // Optional: to verify if Button passes color correctly
        {...props}
      />
    );
  }),
}));

// Import the mocked components to access their Jest mock functions (e.g., for .mockClear()).
import { Icon } from "@components/Atoms/Icons/Icon";
import { Loader } from "@components/Feedback/Loader";
const MockedIcon = Icon as jest.Mock; // Type the mock for better IntelliSense and Jest matchers
const MockedLoader = Loader as jest.Mock; // Type the mock for Loader

describe("@tempo-ui/Atoms/Buttons/button", () => {
  // Clear the mocks before each test to ensure a clean state and prevent test interference.
  beforeEach(() => {
    MockedIcon.mockClear();
    MockedLoader.mockClear();
  });

  describe("with icons", () => {
    it("renders the icon before the label when iconStart & label are present", () => {
      const { container } = render(
        <Button className="customClass" label="label" iconStart={<div className="icon">icon</div>} />,
      );
      // No waitFor needed here, as iconStart is a direct div, not an asynchronous Icon component.
      expect(container.querySelector(".customClass .icon + .tempo-button__label")).toBeInTheDocument();
    });

    it("renders the icon after the label when iconEnd & label are present", () => {
      const { container } = render(
        <Button className="customClass" label="label" iconEnd={<div className="icon">icon</div>} />,
      );
      // No waitFor needed here.
      expect(container.querySelector(".customClass .tempo-button__label + .icon")).toBeInTheDocument();
    });
  });

  describe("enabled", () => {
    it("renders", () => {
      const { asFragment } = render(<Button label="enabled" iconStart={<div>some icon</div>} />);
      // Snapshots should be stable. If the Icon mock always returns the same thing, this is fine.
      // If iconStart were a real Icon, waitFor might be needed.
      expect(asFragment()).toMatchSnapshot();
    });

    it("handles clicks", async () => {
      const clickHandler = jest.fn();
      render(<Button onClick={clickHandler} label="enabled" iconStart={<div>some icon</div>} />);

      // No waitFor needed here unless the button itself renders asynchronously.
      // screen.getByRole("button") will inherently wait a small amount if the element isn't immediately present.
      await userEvent.click(screen.getByRole("button"));
      expect(clickHandler).toHaveBeenCalled();
    });

    describe("renders the dropdown button UI", () => {
      it("without dropdownVisible", async () => {
        render(<Button dropdown label="enabled" />);
        // We need to await waitFor because the Button component renders a mocked Icon,
        // and we want to ensure MockedIcon was called with the correct props.
        await waitFor(() => {
          // Verify that the mocked Icon component was rendered with the correct props.
          // The mock returns an SVG with 'data-icon-name' that we can use.
          expect(screen.getByTestId("mock-icon-arrows-chevron-down")).toBeInTheDocument();
        });
      });

      it("with dropdownVisible", async () => {
        render(<Button dropdown dropdownVisible label="enabled" />);
        await waitFor(() => {
          // Verify that the mocked Icon component was rendered with the correct props.
          expect(screen.getByTestId("mock-icon-arrows-chevron-up")).toBeInTheDocument();
        });
      });
    });
  });

  describe("disabled", () => {
    it("does not call the event handler", async () => {
      const clickHandler = jest.fn();
      render(<Button onClick={clickHandler} disabled />);
      await userEvent.click(screen.getByRole("button"));
      expect(clickHandler).not.toHaveBeenCalled();
    });

    it("renders a loading spinner", async () => {
      const clickHandler = jest.fn();
      render(<Button onClick={clickHandler} loading />); // Render the button in a loading state

      await waitFor(() => {
        // MODIFIED: Now we look for the mocked loader by its data-testid.
        expect(screen.getByTestId("mock-loader")).toBeInTheDocument();
        // Optional: If you want to verify that Button passed the correct class to the mocked Loader.
        expect(screen.getByTestId("mock-loader")).toHaveClass("tempo-button__loader");
      });
      await userEvent.click(screen.getByRole("button"));
      expect(clickHandler).not.toHaveBeenCalled();
    });
  });

  describe("extra", () => {
    it("renders with an additional className", () => {
      render(<Button extra disabled />);
      expect(screen.getByRole("button")).toHaveClass("tempo-button--extra");
    });
  });

  it("adds the additional className", () => {
    render(<Button className="customClass" label="custom" />);
    expect(screen.getByRole("button")).toHaveClass("customClass");
  });

  describe("form", () => {
    it("sets the form attribute if provided", () => {
      render(<Button form="my-form" label="custom" />);
      expect(screen.getByRole("button")).toHaveAttribute("form", "my-form");
    });
  });

  describe("testId prop", () => {
    it("renders the button with the specified data-testid attribute", () => {
      const { getByTestId } = render(<Button label="button-title" testId="button" />);
      const buttonElement = getByTestId("button");
      expect(buttonElement).toBeInTheDocument();
      expect(buttonElement).toHaveAttribute("data-testid", "button");
    });
  });
});
