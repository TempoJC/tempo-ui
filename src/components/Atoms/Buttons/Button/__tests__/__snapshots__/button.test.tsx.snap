// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`@tempo-ui/Atoms/Buttons/button enabled renders 1`] = `
<DocumentFragment>
  <button
    class="tempo-button tempo-button--standard tempo-button--icon-and-label"
    type="button"
  >
    <div
      class="tempo-button__content"
    >
      <div>
        some icon
      </div>
      <span
        class="tempo-button__label"
      >
        enabled
      </span>
    </div>
  </button>
</DocumentFragment>
`;
